quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs

```diff
@@ -515,47 +506,79 @@ runActions
      )
   => Actions state
   -> PropertyM m (Annotated state, Env m)
-runActions (Actions_ rejected (Smart _ actions)) = loop initialAnnotatedState [] actions
+runActions (Actions_ rejected (Smart _ actions)) = do
+  (finalState, env) <- runSteps initialAnnotatedState [] actions
+  unless (null rejected) $
+    monitor $
+      tabulate "Actions rejected by precondition" rejected
+  return (finalState, env)
+
+-- | Core function to execute a sequence of `Step` given some initial `Env`ironment
+-- and `Annotated` state.
+runSteps
+  :: forall state m e
+   . ( StateModel state
+     , RunModel state m
+     , e ~ Error state
+     , forall a. IsPerformResult e a
+     )
+  => Annotated state
+  -> Env m
+  -> [Step state]
+  -> PropertyM m (Annotated state, Env m)
+runSteps s env [] = return (s, reverse env)
+runSteps s env ((v := act) : as) = do
+  pre $ computePrecondition s act
+  ret <- run $ performResultToEither <$> perform (underlyingState s) (polarAction act) (lookUpVar env)
+  let name = show (polarity act) ++ actionName (polarAction act)
+  monitor $ tabulate "Actions" [name]
+  monitor $ tabulate "Action polarity" [show $ polarity act]
+  case (act, ret) of
+    (ActionWithPolarity action PosPolarity, Left err) ->
+      positiveActionFailed action err
+    (ActionWithPolarity action PosPolarity, Right val) -> do
+      (s', env') <- positiveActionSucceeded action ret val
+      runSteps s' env' as
+    (ActionWithPolarity action NegPolarity, _) -> do
+      (s', env') <- negativeActionResult action ret
+      runSteps s' env' as
```

Comment:
This is clearer than before! I like it!

Minor comment: we could do `case (polarity act, ret)` and bind `action = polarAction act` in the `where` clause or in a `let` above to make this a bit less noisy.

Reply:
Yes, that's a good point, I will update the code.